types/jest/index.d.ts

```diff
@@ -282,16 +282,24 @@ declare namespace jest {
      *   spy.mockRestore();
      * });
      */
-    function spyOn<T extends {}, M extends NonFunctionPropertyNames<Required<T>>>(
+    function spyOn<
+        T extends {},
+        M extends keyof T,
+        A extends M extends NonFunctionPropertyNames<Required<T>> ? 'get' | 'set' : never
+            = M extends NonFunctionPropertyNames<Required<T>> ? 'get' | 'set' : never,
+    >(
         object: T,
         method: M,
-        accessType: 'get'
-    ): SpyInstance<Required<T>[M], []>;
-    function spyOn<T extends {}, M extends NonFunctionPropertyNames<Required<T>>>(
-        object: T,
-        method: M,
-        accessType: 'set'
-    ): SpyInstance<void, [Required<T>[M]]>;
+        accessType: A,
+    ): M extends NonFunctionPropertyNames<Required<T>>
+        ? A extends 'set'
+            ? SpyInstance<void, [Required<T>[M]]>
+            : SpyInstance<Required<T>[M], []>
+        : Required<T>[M] extends new (...args: any[]) => any
+        ? SpyInstance<InstanceType<Required<T>[M]>, ConstructorArgsType<Required<T>[M]>>
+        : Required<T>[M] extends (...args: any) => any
+        ? SpyInstance<ReturnType<Required<T>[M]>, ArgsType<Required<T>[M]>>
+        : never;
```

Comment:
Ooff this is rough. Wouldn't it be possible to comment it a bit or make it more maintainable ? Splitting in other generic conditional types ? I don't know if it matches code policies, but it is almost impossible to review.

Reply:
Sure, I'll play around a little to see what I can come up with